### ISSUE TITLE
Compiler should warn if "unused" shadowed variable is actually used in inner block

### ISSUE THREAD

**Issue by tp-m** [rust-lang/rust#87105]
Given the following code: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018

```rust
pub fn main() {
    let mut iter = [1u8, 3, 5, 7].iter().zip([2u8, 4, 6, 8].iter());
    while let Some((a,b)) = iter.next() {
        println!("a: {}, b: {}", a, b);
        if let Some((_a,b)) = iter.next() {
            println!("a: {}, b: {}", a, b);
        }
    }
}
```

The current output is: no warnings.

Ideally the output should look like:

```
Line 5: variable `_a` is marked as unused but shadows variable `a` from line 3.

Did you forget to change `_a` back to `a`?

If not, use `_` instead of `_a` in line 5.
```
Rationale: This might happen if initially `a` from Line 5 wasn't used in the inner block (e.g. it was just `println!("b: {}", b);`). In that case the compiler warns about the unused variable `a` and tells me to change it to `_a` to indicate that it's not used. Equally, the compiler should tell me now if I start using it again and forget to change it back!

**Comment by theo-lw** [comment]
I'd like to try my hand at this @rustbot claim.
